### ISSUE TITLE
LlamaSequenceClassification

### ISSUE THREAD

**Issue by Kleinpenny** [MouYongli/LLMs4OL#4]
我的代码: 
```python

for i, item in enumerate(text):
            item +=  " [SEP] "
            item +=  sentences[i]
            text[i] = item

#{
#    "ID": "__cover_VB_1",
#    "term": "cover",
#   "type": "verb",
#    "sentence": "cover her face with a handkerchief"
#}

#text:   "Given a term: 'cover' [SEP] cover her face with a handkerchief. Classify it into one of type in list
#[’adj‘，’adverb‘，’noun‘，’verb‘]. Classification Result:"  
#label: "verb"

def preprocess_function(examples):
    return tokenizer(examples["term"], truncation=True)

dataset = DatasetDict({'train': Dataset.from_dict({'label': label, 'term': text})})


#Preprocess
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
tokenized_datasets = dataset.map(preprocess_function, batched=True)

#Train
model = AutoModelForSequenceClassification.from_pretrained(
    model_id, num_labels=len(id2label), id2label=id2label, label2id=label2id, device_map="auto"
)
model.config.pad_token_id = model.config.eos_token_id

# Without validation 
training_args = TrainingArguments(
    output_dir=output_dir,
    learning_rate=2e-5,
    per_device_train_batch_size=batch_size,
    num_train_epochs=epochs,
    weight_decay=0.01,
    save_strategy="epoch",
    disable_tqdm=False,
    logging_steps=10,
    save_total_limit=1,
    fp16=True,
    gradient_accumulation_steps=4,
)
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets["train"],
    tokenizer=tokenizer,
    data_collator=data_collator
)

```

关于使用lora的代码:

在加载完模型之后，使用下面的代码:
```python
    if finetune_methode == "lora":
        #Task_Type
        # SEQ_CLS: Text classification.
        # SEQ_2_SEQ_LM: Sequence-to-sequence language modeling.
        # CAUSAL_LM: Causal language modeling.
        # TOKEN_CLS: Token classification.
        # QUESTION_ANS: Question answering.
        # FEATURE_EXTRACTION: Feature extraction. Provides the hidden states which can be used as embeddings or features for downstream tasks.
        peft_config = LoraConfig(
            r=16,
            lora_alpha=32,
            lora_dropout=0.1,
            target_modules=None,  # peft will figure out the correct target_modules based on the model_type
            task_type="SEQ_CLS",
        )

        model = get_peft_model(model, peft_config)
        peft_p = print_number_of_trainable_model_parameters(model)
        print(f"# Trainable Parameter \nBefore: {ori_p} \nAfter: {peft_p} \nPercentage: {round(peft_p / ori_p * 100, 2)}")
        output_dir += "lora"
```

**Comment by Kleinpenny** [comment]
```python
class transformers.LlamaForSequenceClassification:
```
该类包含一个序列分类头部（线性层）用于 LLaMa 模型，并使用最后一个token来进行分类
如果配置中定义了 pad_token_id，模型会找到每行中最后一个非填充token。如果没有定义 pad_token_id，则直接使用每行的最后一个token。

```python
init()
```

self.score：定义一个线性层（nn.Linear），用于将模型的隐藏状态（大小为 hidden_size）映射到分类标签数量上，用于最终分类。-> 多标签，CrossEntropy

每个词的embedding经过多层自注意力和前馈网络处理，逐层生成一系列hidden states。最后一层的hidden states表示该词的最终上下文表示。hidden states: (batch_size, sequence_length, hidden_size)

```python
forward()
```
input_ids : (batch_size, sequence_length)

**Comment by Kleinpenny** [comment]
我的代码:

- 每个data: "term" + [SEP] + "sentence"
- preprocess_function 函数对输入的 term 字段进行分词并截断，确保输入长度符合模型的限制
- pad_token 设置为 eos_token，即填充标记为结束标记, 

### Trainer
- **forward()**：生成(4, num_of_lables)的logits,
- **loss calculation**：`Trainer` 使用交叉熵损失函数。

![image](https://github.com/user-attachments/assets/16557efe-7fc9-480c-9eae-1de295285a24)
𝑦𝑖是真实标签，𝑝𝑖是模型预测的概率分布
  
- **Backpropagation**：基于计算出的损失值，`Trainer` 使用反向传播（Backpropagation）算法计算损失对每个参数的梯度。
- **Parameter Update**：通过优化器（默认是 AdamW）使用梯度更新模型的权重，使损失降低。更新频率根据 `gradient_accumulation_steps = 4`控制，因此每 4 个steps会被累积后再更新一次权重。
- 
在 `TrainingArguments` 中设置了 `logging_steps=10`，`Trainer` 会计算 10 个stpes（共 40 个批次）的平均损失值

CrossEntropy的计算例子:
当 `num_labels=4` 时，这意味着这是一个四分类问题。模型在计算损失时会使用 **交叉熵损失**（Cross-Entropy Loss），适用于多类别分类任务。以下是详细的损失计算过程：

### 交叉熵损失的计算步骤

1. **Logits 输出**：
   - 模型的 `forward` 方法会输出形状为 `(batch_size, num_labels)` 的 logits 张量。例如，假设 `batch_size=4`，那么 logits 的形状将为 `(4, 4)`。
   - 每个样本的 `logits` 向量包含 4 个值，表示该样本在每个类别上的非标准化预测得分。

2. **Softmax 转换**：
   - 在交叉熵损失计算之前，`logits` 会先通过 **softmax** 转换成概率分布。Softmax 会将 logits 转换为一个范围在 `[0, 1]` 的概率分布，其中每个样本的所有类别概率之和为 1。
   - 公式为：
![image](https://github.com/user-attachments/assets/90a351e7-674c-466b-906f-de30f1087094)
3. **交叉熵损失计算**：
   - 将转换后的概率分布与实际标签（`labels`）进行比对，计算交叉熵损失。对于每个样本，如果其真实标签为 `y`，交叉熵损失的公式为：
![image](https://github.com/user-attachments/assets/a544f626-6b2d-4136-a91b-163d03a88b7f)
   - 对于一个批次中的所有样本，计算所有样本的损失并取平均值：

![image](https://github.com/user-attachments/assets/328a0af3-8c37-45d9-9978-113ff498d08a)

   其中 \( p_{i, y_i} \) 是第 \( i \) 个样本的真实类别的预测概率。

**Comment by Kleinpenny** [comment]
### Lora

只对 LoRA 添加的低秩矩阵部分进行训练，而保持模型其他参数冻结不变

确定任务类型L SEQ_CLS,那么会选择LLaMa的attention和feed_forward层，特别是：
- 自注意力层中的投影矩阵（例如 q_proj, v_proj, k_proj）：在这些层上加入 LoRA 的低秩矩阵，以便调整注意力机制中关键的投影。
- 前馈网络层中的特定参数矩阵（如 intermediate_dense 和 output_dense）：可以增强特征表示的选择性，从而更好地支持分类决策。
